Docs: improve accuracy of sqlite3 `check_same_thread` parameter (#101351)
python/cpython: Doc/library/sqlite3.rst

   :type isolation_level: str | None

   :param bool check_same_thread:

>        If ``True`` (default), an exception will be raised if the database connection

Reviewer: ```suggestion
       If ``True`` (default), :exc:`ProgrammingError` will be raised if the database connection
```

Reviewer: (I'll reflow before merging, FYI)

Author: accepting as is, since it is already better than current text

>        is used by a thread other than the one that created it.

Reviewer 2: Can we be more precise about what "used" means here? Simply if the top-level object is accessed in any manner? Calling specific read and/or write methods?

Reviewer: Well, the SQLite docs use similar wording and is just as vague (https://sqlite.org/threadsafe.html).

_I_ assume this means usage at the data structure level, so for example for the _SQLite multi-threaded_ mode, I assume all C APIs that work with `sqlite3 *` pointers are _not_ thread safe, but all C APIs that work with `sqlite3_stmt *` pointers are thread safe. This is hard to translate to Python APIs, though, as there is no one-to-one mapping of sqlite3 Python APIs and SQLite C APIs.

For example, most of the Python APIs of `sqlite3.Cursor` map to SQLite C APIs that deal with `sqlite3_stmt *`; this is fine... until an exception is raised! All exceptions raised by the sqlite3 module use C APIs that access the current connection. Thread safe? I don't know; perhaps you'll end up with the wrong error message? Or perhaps you'll end up with no error message at all, in case the sqlite3 module will return `NULL` from a Py\_ C API _without_ an exception being set, which is a fatal error; you'll get a seg fault, IIRC.

If the underlying SQLite module is using **serialized** mode, you can of course safely set `check_same_thread` to `False`.

       This may be helpful in particular when SQLite is not in serialized mode

>        (see :attr:`threadsafety`).

Reviewer: Hm, I find this slightly more difficult to read than the original. This is a reference, so it should be as accurate and to the point as possible (please see [^1]). IMO, _"If False"_ is a better way to start this sentence because it is unambiguous and succinct.

I'm afraid I don't have a useful suggestion to come up with today (I'll sleep on it).

[^1]: https://diataxis.fr/reference/

Reviewer 2: Yeah, IMO this line is a significant regression on the previous revision for unclear benefit, removing explicit description of what `False` does as well as other relevant and appropriate information, adopting a tone not consistent with use in a reference, and overall much less clear and more confusing to me at least as a non-expert. What's more, I'm not even sure this directly address @erlend-aasland 's original comment that prompted this. I'm sorry if I confused you with my feedback, as this was not at all what I suggesting.

I would suggest just reverting to the previous wording (with the "please see" fix), and considering if any of the ≈dozen much more focused suggestions offered previously, or some variant of the same, resolve the concern around "seriali

Here's that, as a baseline to start with, with a few much more conservative fixes and tweaks to start off with:

```suggestion
       If ``False``, the connection may be accessed in multiple threads;
       write operations may need to be serialized by the user
       to avoid data corruption.
       See :attr:`threadsafety` for more information.
```

   :param bool check_same_thread:
       If ``True`` (default), only the creating thread may use the connection.
       If ``False``, the connection may be shared across multiple threads;

>        users should check :attr:`threadsafety` before setting this value to
>        avoid data corruption.

Reviewer: I worry that the suggested text may come across as too vague. Parallell programming is hard to get right, even for pros. I'd be more comfortable if we keep the current serialisation recommendation, and instead _add_ a sentence where we point to the threadsafety attribute.

How about this:

```suggestion
       if so, write operations may need to be serialized by the user,
       to avoid data corruption.
       Please see :attr:`threadsafety` for more information.
```

   :param bool check_same_thread:
       If ``True`` (default), only the creating thread may use the connection.
       If ``False``, the connection may be shared across multiple threads;

>        if so, write operations may need to be serialized by the user,

Reviewer 2: What about thinking outsize the "ized" and stating more explicitly what the user needs to do, instead of how they should do it?

```suggestion
       if so, write operations must only be performed in series,
```

or

```suggestion
       if so, write operations must only be performed serially,
```

or

```suggestion
       if so, the user should ensure write operations are only performed serially,
```

Or what about instead flipping it and talking about what the user needs to ensure _doesn't_ happen?

```suggestion
       if so, write operations must not performed in parallel,
```

or 

```suggestion
       if so, the user should ensure write operations are not performed in parallel,
```

Reviewer: Marking as resolved (outdated)

       to avoid data corruption.

>        Please see :attr:`threadsafety` for more information.

Reviewer 2: IMO, "Please" changes the meaning of this sentence, from simply informing the user they can consult another resource for more information, to actively begging them to do so. I'm not sure the latter is necessary here, nor particularly consistent with the general tone of these reference docs.

```suggestion
       See :attr:`threadsafety` for more information.
```
